utils/chunked_vector.hh

```diff
@@ -325,8 +327,25 @@ size_t chunked_vector<T, max_contiguous_allocation>::external_memory_usage() con
 template <typename T, size_t max_contiguous_allocation>
 chunked_vector<T, max_contiguous_allocation>::chunked_vector(const chunked_vector& x)
         : chunked_vector() {
-    reserve(x.size());
-    std::copy(x.begin(), x.end(), std::back_inserter(*this));
+    auto size = x.size();
+    if (!size) {
+        return;
+    }
+    reserve(size);
+    auto copy_chunk = [this, &x] (size_t i, size_t n) {
+        const T* src = x._chunks[i].get();
+        T* dst = _chunks[i].get();
+        std::uninitialized_copy_n(src, n, dst);
+    };
+    size_t i;
+    size_t full_chunks = size / max_chunk_capacity();
+    for (i = 0; i < full_chunks; ++i) {
+        copy_chunk(i, max_chunk_capacity());
+    }
+    if (auto remainder = size % max_chunk_capacity()) {
+        copy_chunk(i, remainder);
+    }
```

Comment:
The if() is likely unneeded. But maybe it is - can _chunks be empty?

Reply:
yes, _chunks can be empty if the chunked_vector is empty.
Also _size might be aligned on `max_chunk_capacity()` and then there is no remainder.
We can turn this into an iteration on _chunks.size(), calculating size for chunk for each iteration.
It might be a bit more efficient for single chunk vectors

Reply:
Then copy_chunk(0) will be illegal.

Reply:
That's why I had the first `if (!size)` in the beginning of the function.
In any case, godbolt .org indicates that a single loop will produce slightly more condensed machine code for x86_64, so I simplified the function in v3, by having a single loop and calculating the count per chunk in the loop body using std::min(size, max_chunk_size())

Reply:
Yes the new loop looks much better.

Reply:
But: it's not exception safe. You must update _size after every uninitialized_copy_n() succeeds, so the destructor can tear it down if the next one fails (it will destroy partial work that it does, not not work that previous iterations did).

Reply:
No, you need a try/catch, since the destructor won't be called on a not-fully-constructed object.

Reply:
good point, will fix.

Reply:
Actually, @avikivity, the copy constructor calls the default constructor (which is implicitly noexcept).
So the destructor will called if an exception is thrown in the copy constructor function body, since it's already considered as fully constructed.

Reply:
In the process, I've found out that the fill constructor isn't safe.  Submitted https://github.com/scylladb/scylladb/pull/18636 (separately, to consider backport)